### DISCUSSION TITLE
Geräte Firmware Aktualisierungen über HA

### DISCUSSION THREAD

**Discussion by SukramJ** [SukramJ/aiohomematic#1061] [Show and tell]
Mit der Version 1.37.0 können die Geräte Firmware Installationen über HA gestartet werden.
Weitere Informationen findet ihr in der [README](https://github.com/danielperna84/custom_homematic#updating-a-device-firmware).

**Comment by emufan** [comment]
Sehr fein. Werde probieren.

Bislang zeigt er mir noch kein Update an, auch wenn es welche gibt. Allerdings sollen die auch gar nicht installiert werden (und ich lade sie daher auch nicht auf die CCU), da es Probleme damit geben soll. 

Also werde ich ebenfalls probieren, diese zu skippen, falls sie auftauchen.

**Reply by SukramJ** [reply]
Falle es nicht klar wurde: Die Datei muss nach wie vor zur CCU hochgeladen werden.

**Reply by emufan** [reply]
Doch hatte ich so verstanden. Dachte aber dass er ein Vorhandensein eines Updates aus zeigt, bzw. Du das auch mit den alle 6 Stunden meintest, d.h. analog

![image](https://github.com/danielperna84/hahomematic/assets/22775515/177af7f1-62fc-4e28-b014-b427c1eab1c1)

Ist/wäre für mich aber alles nicht kriegsentscheidend.

**Comment by noxhirsch** [comment]
Hi @SukramJ,
vielen Dank für die Implementierung der Updates - das ist echt super praktisch! Einen kleinen Änderungsvorschlag/-wunsch hätte ich:
Während das Update übertragen wird, sollten IMHO die Update-Sensoren noch nicht den Status "installing" haben. Dadurch ist das Update (bzw. viele Updates, wenn man mehrere gleiche Geräte / mehrere Updates hat) dauerhaft und nicht entfernbar auffällig in der Sidebar und der Update-Sektion sichtbar - und das je nach Anzahl der Updates tage- bis wochenlang.
![image](https://github.com/danielperna84/hahomematic/assets/30938717/2cd08f14-84c4-4151-9a08-ccb0d46407d1)
Ich würde vorschlagen, dass der Vorgang der Update-Übertragung nicht im Frontend in den Updates-Entities sichtbar gemacht werden sollte, sodass diese erst in der Update-Sektion auftauchen, wenn das Update tatsächlich bereit zur Installation ist.

**Reply by SukramJ** [reply]
Das könnte man auch so machen.

**Reply by emufan** [reply]
Nachdem hier aktuell auch viele anstehen und noch viele folgen werden, da sie ja aktuell mehr als viele Updates veröffentlichen, fände ich es auch gut, wenn sie erst erscheinen, wenn übertragen wurden. Denn das zieht sich ja meist wirklich ewig.

**Reply by SukramJ** [reply]
Ganz so einfach ist das dann doch nicht.

Mögliche HM firmware [Stati](https://github.com/danielperna84/hahomematic/blob/b64567a3129bc78b61594ace3ad8ddf8062abf2e/hahomematic/const.py#L352-L360) 

Sobald eine Firmware Update Datei in der CCU hochgeladen wurde ändert sich die `latest_version`. Wenn diese grösser ist als die `installed_version` wird der Update-Button angezeigt. Bei HM wird jetzt aber erst die Datei auf das Gerät übertragen, so dass der Update-Button noch keinen Sinn macht.
Damit der Update-Button nicht erscheint kann man das HA property `in_progress` auf `true` setzten.
Ich hab das jetzt so gewählt, das `in_progress` gleich `true` ist wenn das Gerät **nicht** den Firmwarestatus `UP_TO_DATE` oder `READY_FOR_UPDATE` hat.

Wenn Ihr eine bessere Idee habt, dann raus damit. An der `latest_version` wollte ich eigentlich nicht rum manipulieren.

**Reply by emufan** [reply]
Hört sich doch schlüssig an. 👍

**Reply by SukramJ** [reply]
Danke, deswegen ist das ja der Ist-Zustand.

**Reply by emufan** [reply]
Hm, dachte das wäre die Änderung und nicht der Ist-Zustand. Dann habe ich es ansonsten falsch verstanden. Ich dachte Dein Vorschlag wäre nun:

Setzte zeige das Update mitsamt Update-Button generell nur an, wenn
(('''latest_version''' > '''installed_version'')' **und** (wenn das Gerät nicht den Firmwarestatus '''UP_TO_DATE''' oder '''READY_FOR_UPDATE''' hat)).

**Reply by SukramJ** [reply]
Die HA Regel ist ja nur ('''latest_version''' > '''installed_version'')
Das ist in der Integration nicht überschreibbar.

**Reply by SukramJ** [reply]
Daher ist in_progress ja auch inkl. FW delivery.

**Reply by emufan** [reply]
Ach so. Ich dachte Du bekommst die Status, d.h. auch ```latest_version``` und ```installed_version``` und lieferst den Update-verfügbar selbst an HA. O.k. wenn dem nicht so ist, dann habe ich aktuell auch keine bessere Idee, außer das Manipulieren der latest_version, was aber sicher nicht so fein ist.

**Reply by noxhirsch** [reply]
> Sobald eine Firmware Update Datei in der CCU hochgeladen wurde ändert sich die latest_version. Wenn diese grösser ist als die installed_version wird der Update-Button angezeigt.

> Das ist in der Integration nicht überschreibbar.

Ich nehme an, dass dieser Vergleich den Status (on/off) der Update Entity festlegt und du eben genau diesen nicht ändern kannst, oder? Ich habe diesen testweise über die Developer Tools auf "off" gesetzt und das Update blieb dann verschwunden, bis es installationsbereit war. 
Für mich persönlich, wäre das ein Workaround, dass ich über einen Automation den Status überschreibe - aber ich denke nicht, dass für alle Nutzer sinnvoll ist.
Ich persönlich fänd es allerdings auch keine schlechte Lösung, die `latest_version` zu überschreiben. Die beschreibt ja, welche Version zur Installation bereit ist - und bis zu dem Zeitpunkt, zu dem die neue Firmware übertragen ist, ist ja tatsächlich noch keine neuere Version installationsbereit ;)

**Reply by emufan** [reply]
> Für mich persönlich, wäre das ein Workaround, dass ich über einen Automation den Status überschreibe - aber ich denke nicht, dass für alle Nutzer sinnvoll ist.

Wie würdest Du das machen? Habe da keinen funktionierenden Service gefunden im HA-Standard. Es geht selbst für diese update-Domäne nicht mit dem sonst benutzten Helpern aus der HACS-Variables-Integration, mit dem man auch sonst sensoren usw. manipulieren kann.

Im Moment habe ich tatsächlich alle update-Entitäten der Integration wieder deaktiviert. Aktuell laufen bei mir 2 Gerätetetypen mit 10 Geräten und etliche Typen mit zusammen über 60 Geräten folgen danach. Da hätte ich nun sicher über Wochen oder länger 10-15 ständige Zahlen unten im HA ohne was machen zu können.

**Reply by noxhirsch** [reply]
Ich habe dafür ein Python Script, das ich mal vor Jahren im Forum gefunden habe:
<details>
  <summary>Script</summary>
  
`/config/python_scripts/set_state.py`
```
if 'entity_id' not in data:
    logger.warning("===== entity_id is required if you want to set something.")
else:
    data = data.copy()
    inputEntity = data.pop('entity_id')
    inputStateObject = hass.states.get(inputEntity)
    if inputStateObject:
        inputState = inputStateObject.state
        inputAttributesObject = inputStateObject.attributes.copy()
    else:
        inputState = 'unknown'
        inputAttributesObject = {}
    if 'state' in data:
        inputState = data.pop('state')
    logger.debug("===== new attrs: {}".format(data))
    inputAttributesObject.update(data)

    hass.states.set(inputEntity, inputState, inputAttributesObject)
```

Im service call kann man dann definieren, was überschrieben werden soll:
```
service: python_script.set_state
data:
  entity_id: update.az_desk_button_update
  state: off
```
  
</details>

**Reply by emufan** [reply]
> Ich habe dafür ein Python Script, das ich mal vor Jahren im Forum gefunden habe:

Prima. Geht. Danke!

Dann gucke ich mal ob die die Update-Entitäten doch die Tage mal wieder aktiviere und mir auch eine Automatisierung über alle dieser baue, mit setze off wenn ```firmware_update_state: DELIVER_FIRMWARE_IMAGE``` oder ```in_progress: true``` (muss mal gucken wie sich ```firmware_update_state``` ändert, wenn mal ein Ding durch ist, hatte bisher nicht drauf geguckt), was on change von den Entitäten und bei HA-Restart (da gehen sie wieder on) auf off setzt.  

Inkl. des Skripts, usw. aber tatsächlich gefühlt mehr Hack als evtl. dich ein Manipulieren der ```latest_version```. Aber ich kann auch verstehen, wenn SukramJ das nicht gut findet und daher nicht machen würde/wird.

Update: Wird noch etwas komplexer die Automation. Man muss wirklich HA-restart nochmal nehmen, da eine Update-Entity, die vorher ausgeschaltet wurde und beim Neustart wieder an geht, keinen state-trigger-on triggert. Also mind. einmal beim Restart nochmal in einen For-Each-Loop und nochmal alle prüfen und ggf. wieder ausmachen. Und zusätzlich ja nicht nur auf on triggern, sondern eher (zusätzlich oder nur) auf die Attribute, da es ja auch danach on bleibt, wenn ```in_progress``` oder ```firmware_update_state``` sich ändert und das ja quasi der Filter ist. Also was für ein langes Wochenende. ;-)

**Reply by SukramJ** [reply]
Alternativ könntet ihr auch die 1.37.1 testen. 
Da werden die Updates erst nach der Übertragung angezeigt.
`in_progress` wird nur noch angezeigt wenn das Update ausgeführt bzw. auf die Update Ausführung gewartet(Batterie betrieben Geräte) wird.

**Reply by emufan** [reply]
> Alternativ könntet ihr auch die 1.37.1 testen. Da werden die Updates erst nach der Übertragung angezeigt. `in_progress` wird nur noch angezeigt wenn das Update ausgeführt bzw. auf die Update Ausführung gewartet(Batterie betrieben Geräte) wird.

Kaum schreibe ich mein Update eben, bist Du schon schneller mit einer neuen Version. 😂👍

Aus Interesse: Wie hast Du es nun gemacht/was geändert?

**Reply by SukramJ** [reply]
> Aus Interesse: Wie hast Du es nun gemacht/was geändert?

1. `latest_version` wird erst angezeigt, wenn FW-Datei auf Gerät übertragen wurde (`READY_FOR_UPDATE`)
2. `in_progress` wird angezeigt, wenn Gerät im Zustand `DO_UPDATE_PENDING` oder `PERFORMING_UPDATE`

**Reply by SukramJ** [reply]
Wichtig: Bitte Rückmeldung, ob das jetzt auch so klappt. Bei mir sind alle Updates durchgelaufen.

**Reply by noxhirsch** [reply]
Vielen Dank! Das Testen muss ich aber @emufan überlassen - ich habe aktuell auch keine ausstehenden Updates.

**Reply by emufan** [reply]
> Wichtig: Bitte Rückmeldung, ob das jetzt auch so klappt. Bei mir sind alle Updates durchgelaufen.

Also weg sind sie aktuell erst mal und Status ist z.B. dieser:

![image](https://github.com/danielperna84/hahomematic/assets/22775515/3c3b9e17-6f02-478a-b659-1fd82e34b17f)

Muss dann mal gucken, ob sie wiederkommen, wenn übertragen und werde berichten ...

> Vielen Dank! Das Testen muss ich aber @emufan überlassen - ich habe aktuell auch keine ausstehenden Updates.

Wie macht ihr das? Bis meine noch anstehenden durch sind, wird es noch Wochen brauchen. :-( Aber klar, liegt sicher an meinen Typen und deren jeweiliger Anzahl.

**Reply by noxhirsch** [reply]
> Bis meine noch anstehenden durch sind, wird es noch Wochen brauchen.

Theoretisch müssten die Update Entities jetzt nach und nach auftauchen. Die Updates werden vermutlich mehr oder weniger nacheinander auf die Geräte übertragen - zumindest war das bei mir immer so. Da waren dann immer alle paar Tage ein-zwei neue Geräte updatebereit.

**Reply by emufan** [reply]
> > Bis meine noch anstehenden durch sind, wird es noch Wochen brauchen.
> 
> Theoretisch müssten die Update Entities jetzt nach und nach auftauchen. Die Updates werden vermutlich mehr oder weniger nacheinander auf die Geräte übertragen - zumindest war das bei mir immer so. Da waren dann immer alle paar Tage ein-zwei neue Geräte updatebereit.

Ja klar. Aber bei der letzten Update-Orgie von eq3 gab es bei mir so viele, bis die übertragen sind ... Und damit hier nicht zuviel übertragen wird, mache ich es eh meist Typ für Typ. Aktuell laufen bei mir SRH und STHD. Aber es stehen noch aus SWD, BROLL (wobei die ja Fehlerhaft ist und ich sie erst mal skippe), WTH-2, SMI, BWTH, ...

**Reply by noxhirsch** [reply]
Achso, jetzt verstehe ich die Frage. Ich lade die verfügbaren Updates eigentlich alle auf einmal rein. Die CCU regelt das eigentlich ganz gut, dass da nicht zu viel auf einmal übertragen wird und der Duty Cycle im grünen Bereich ist (das hängt aber sicherlich auch davon ab, wie viel Puffer man da sonst hat). Danach checke ich täglich / alle paar Tage und starte die Installationen - und das wird ja jetzt praktischerweise deutlich einfacher.
Das hat bei mir bisher keine Probleme gegeben - ich hatte bis vor kurzem 84 Geräte (12 Bewegungsmelder habe ich gerade entfernt) aber aus "historischen Gründen" auch noch 2 HAPs (neben der CCU).

**Reply by emufan** [reply]
> Muss dann mal gucken, ob sie wiederkommen, wenn übertragen und werde berichten ...

So. Erstes Update ist wieder eingetrudelt. 

![image](https://github.com/danielperna84/hahomematic/assets/22775515/8d2dd9b9-4128-439c-bacf-774c2a73044c)

Scheint also zu klappen wie nun neu designed und wie ich es auch besser finde.

Und ging dann auch über Zwischenstatus ```DO_UPDATE_PENDING``` wieder weg. Prima!

**Comment by emufan** [comment]
Exkludierst Du wie in in anderen Fällen auch irgendwo, wo der Update-Button nicht relevant ist? Wenn ja, mir sind die hier bisher aufgefallen:

HmIP-RCV-50 HmIP-RCV-1
HM-RCV-50 BidCoS-RF

Stört dort nicht, aber hängen ja eigentlich schon und allein an der CCU. Wenn nein, egal.

OffTopic: Woher kommt die FW-Version an dem CCU-Gerät, an dem **nicht** die SVs hängen?

![image](https://github.com/danielperna84/hahomematic/assets/22775515/14883258-58bf-40e5-a91d-50a3d40bfb12)

vs dem, wo die SVs liegen

![image](https://github.com/danielperna84/hahomematic/assets/22775515/568f0e7e-6566-4822-9582-f713c81dfc62)

**Reply by SukramJ** [reply]
> HmIP-RCV-50 HmIP-RCV-1
HM-RCV-50 BidCoS-RF

An die hab ich gar nicht gedacht. Danke für den Hinweis.


Bild 1 ist das Funkmodul, das wie ein normales Gerät ausgelesen wird. Der `device_type` HmIP-CCU3 kommt auch vom Gerät.

Bild 2 ist die CCU. Hier wird die FW Version über eine andere API gelesen. Der `device_type` CCU wird manuell vergeben.
